javalib/src/main/scala/java/math/Multiplication.scala

```diff
@@ -0,0 +1,454 @@
+ */
+object Multiplication {
+
+  /**
+   * An array with powers of ten that fit in the type {@code int}.
+   * ({@code 10^0,10^1,...,10^9})
+   */
+  val tenPows = newArrayOfPows(10, 10)
+
+  /**
+   * An array with powers of five that fit in the type {@code int}.
+   * ({@code 5^0,5^1,...,5^13})
+   */
+  val fivePows = newArrayOfPows(14, 5)
+
+  /**
+   * An array with the first powers of ten in {@code BigInteger} version.
+   * ({@code 10^0,10^1,...,10^31})
+   */
+  val bigTenPows = new Array[BigInteger](32)
+
+  /**
+   * An array with the first powers of five in {@code BigInteger} version.
+   * ({@code 5^0,5^1,...,5^31})
+   */
+  val bigFivePows =  new Array[BigInteger](32)
+
+  val whenUseKaratsuba = 63
+
+  initialiseArrays()
+
+  /**
+   * Multiplies an array of integers by an integer value.
+   *
+   * @param a the array of integers
+   * @param aSize the number of elements of intArray to be multiplied
+   * @param factor the multiplier
+   * @return the top digit of production
+   */
+  def multiplyByInt(a: Array[Int], aSize: Int, factor: Int): Int =
+    multiplyByInt(a, a, aSize, factor)
+
+  /**
+   * Multiplies a number by a positive integer.
+   * @param bi an arbitrary {@code BigInteger}
+   * @param factor a positive {@code int} number
+   * @return {@code val * factor}
+   */
+  def multiplyByPosInt(bi: BigInteger, factor: Int): BigInteger = {
+    val resSign: Int = bi.sign
+    val aNumberLength = bi.numberLength
+    val aDigits = bi.digits
+
+    if (resSign == 0) {
+      BigInteger.ZERO
+    } else if (aNumberLength == 1) {
+      val res: Long = unsignedMultAddAdd(aDigits(0), factor, 0, 0)
+      val resLo = res.toInt
+      val resHi = (res >>> 32).toInt
+      if (resHi == 0) new BigInteger(resSign, resLo)
+      else new BigInteger(resSign, 2, Array(resLo, resHi))
+    } else {
+      val resLength = aNumberLength + 1
+      val resDigits = new Array[Int](resLength)
+      resDigits(aNumberLength) = multiplyByInt(resDigits, aDigits, aNumberLength, factor)
+      val result = new BigInteger(resSign, resLength, resDigits)
+      result.cutOffLeadingZeroes()
+      result
+    }
+  }
+
+  /**
+   * Multiplies a number by a power of ten.
+   * This method is used in {@code BigDecimal} class.
+   * @param bi the number to be multiplied
+   * @param exp a positive {@code long} exponent
+   * @return {@code val * 10<sup>exp</sup>}
+   */
+  def multiplyByTenPow(bi: BigInteger, exp: Long): BigInteger = {
+    if (exp < tenPows.length) multiplyByPosInt(bi, tenPows(exp.toInt))
+    else bi.multiply(powerOf10(exp))
+  }
+
+  /**
+   * Performs a<sup>2</sup>.
+   *
+   * @param a The number to square.
+   * @param aLen The length of the number to square.
+   */
+  def square(a: Array[Int], aLen: Int, res: Array[Int]): Array[Int] = {
+    var carry: Long = 0l
+
+    for (i <- 0 until aLen) {
+      carry = 0
+      for (j <- i + 1 until aLen) {
+        carry = unsignedMultAddAdd(a(i), a(j), res(i + j), carry.toInt)
```

Comment:
Use another `val` with a different name than `carry` here. At this point what you assign there is not a carry at all, so this is misleading.
You can then declare `carry` itself as an `Int` instead of a `Long` in the first place.